### PR TITLE
parser: Remove support for inner attributes on non-block expressions

### PR DESCRIPTION
**Pull request by petrochenkov**
Remove support for attributes like
```rust
fn attrs() {
    (#![print_target_and_args(fifth)] 1, 2);

    [#![print_target_and_args(sixth)] 1 , 2];
    [#![print_target_and_args(seventh)] true ; 5];


    match 0 {
        #![print_target_and_args(eighth)]
        _ => {}
    }

    MyStruct { #![print_target_and_args(ninth)] field: true };
}
```
They are
- useless
- unstable (modulo holes like https://github.com/rust-lang/rust/issues/65860)
- pessimize compiler performance, namely token collection for macros (cc https://github.com/rust-lang/rust/pull/82608)

I still want to run crater on this to check whether the stability holes are exploited in practice, and whether such attributes are used at all.

### CODE REVIEW

**Comment by petrochenkov** [comment]
@craterbot check

**Comment by petrochenkov** [comment]
4 unique legitimate regressions:
- `match port { #![allow(clippy::cast_ptr_alignment)] ... }`
- ``match self.color { //! The `format!()` [macro] lets us create a `String` with a pattern, ... }``
- `match *self { #![allow(unknown_lints,match_same_arms)] ... }`
- `match mdid { #![allow(clippy::unreadable_literal)] }`

**Comment by petrochenkov** [comment]
## FCP report

I propose removing support for inner attributes on tuple, parentheses, array, struct and match expressions from the language.

Examples:
```rust
fn attrs() {
    (#![print_target_and_args(fifth)] 1, 2);

    [#![print_target_and_args(sixth)] 1 , 2];
    [#![print_target_and_args(seventh)] true ; 5];


    match 0 {
        #![print_target_and_args(eighth)]
        _ => {}
    }

    MyStruct { #![print_target_and_args(ninth)] field: true };
}
```

Such attributes are not used in practice and pose challenges to token collection performed during parsing for various macros to work correctly (https://github.com/rust-lang/rust/pull/82608).
Expressions are the most common AST nodes and we have to collect tokens for most of them if we assume that they can potentially contain inner attributes, pessimizing the compiler significantly.

Support for these attributes is a part of `#![feature(stmt_expr_attributes)]`.
Attributes on expressions and statements, including inner attributes, were introduced by RFC 16 https://github.com/rust-lang/rfcs/blob/master/text/0016-more-attributes.md.
The RFC, however says "Inner attributes can be placed at the top of blocks (and other structure incorporating a block) and apply to that block." (https://github.com/rust-lang/rfcs/blob/master/text/0016-more-attributes.md#inner-attributes), which doesn't apply to any cases removed in this PR.
The RFC gives an example with `match` though, which is not a block.
Others cases (tuples, etc.) were added in https://github.com/rust-lang/rust/pull/29850 without any discussion.

So, removal of these cases also makes the rule for "where are inner attributes accepted" much simpler and aligns it with the original RFC.

### Detailed design

Inner attributes are now accepted only on crate/module files and braced blocks (`{ ... }`) containing statements (or some subset of statements like items).
Support for inner attributes in other places is removed.

### Issues

Feature gating for attributes on expressions and statements is exceptionally random and buggy.
As a result some of the inner attributes in question are also available on stable in some cases (the attribute must be a built-in, and the expression must be an expression statement).
Fortunately, we are saved here by the fact that these attributes are not useful in practice, crater found only 4 uses of such attributes, all on `match` (https://github.com/rust-lang/rust/pull/83312#issuecomment-809045818, https://github.com/rust-lang/rust/pull/83312#issuecomment-809599116).

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

I'm going go ahead and kick off an FCP. This is also nominated for discussion in today's lang-team meeting.

**Comment by Aaron1011** [comment]
Note that in the final approach taken in https://github.com/rust-lang/rust/pull/82608, inner attributes on expressions are now less of a problem than they once were. Custom inner attributes in expressions are banned, so we only need to pessimize token collection in the presence of `#[derive]` or `#[cfg_attr]`. However, there are still performance wins to be had in these cases, and the performance issue would likely become significant again if we decided to perform cfg-stripping for all attribute macros (in addition to derive macros).

File: src/test/ui/proc-macro/inner-attrs.rs [hunk 1]
```diff
@@ -34,8 +35,6 @@ struct MyStruct {
 struct MyDerivePrint {
     field: [u8; {
         match true {
-            #![cfg_attr(not(FALSE), rustc_dummy(first))]
```
Comment (Aaron1011) [inline src/test/ui/proc-macro/inner-attrs.rs:37]: Do we have a test that checks that this now errors?

**Comment by Aaron1011** [comment]
r=me with the comment addressed

**Comment by nikomatsakis** [comment]
Uh, @rust-lang/lang, I kind of missed the boat here -- I had an action item to leave a concern about the backwards compatibility angle here, [as we discussed at our last meeting](https://github.com/rust-lang/lang-team/blob/master/minutes/2021-04-27.md#parser-remove-support-for-inner-attributes-on-non-block-expressions-rust83312), but I fell behind and didn't realize this was time sensitive. 

I guess I will open an issue to discuss.

**Comment by nikomatsakis** [comment]
Opened https://github.com/rust-lang/rust/issues/84879

**Pull request merged**